doc/reference/trd104-syscalls.md

```diff
@@ -133,7 +133,55 @@ to `self` as their first argument. Therefore, `r0` is usually used to dispatch
 onto the correct driver; this argument is consumed by the system call handler
 and replaced with `&self` when the actual system call method is invoked.
 
-3.2 Return Values
+3.2 Syscall Argument Types
+----------------------------------
+
+The use of registers, and the selection of specific registers, imposes physical
+restrictions on the types which can be passed (i.e., this TRD defines the ABI
+for platforms with 32-bit, general purpose registers, which limits any single
+argument to a maximum of 32-bit storage for its type).
+
+Syscalls SHOULD define as specific type as is semantically meaningful for these
+32 bits of storage. For example, the `command` syscall uses register 0 as a
+Driver Number and register 1 as a Command Number. These are defined as a `u32`,
+as Tock expects no more than 2^32 drivers, each with no more than 2^32
+commands. A future 64-bit ABI MAY choose to combine these into one syscall
+register. In contrast, other syscall argument slots may have more generic
+purposes, where the type and underlying storage are likely to move in sync on
+future platforms. This document defines the following conceptual types:
+
+ - `OPAQUE_NUMERIC`: This describes an argument akin to C's `(unsigned) int`,
+   i.e., a whole number which is capable of being expressed in a native machine
+   word type.
+ - `[OPAQUE_]SIZE`: This describes an argument akin to C's `size_t`, i.e.,
+   an unsigned, numeric type with a range capable of expressing in bytes the
+   size of any valid _contiguous_ object.
+ - `[OPAQUE_]POINTER_OR_ZERO`: This describes an argument similar to C's
+   `uintptr_t`, i.e., a type capable of holding a pointer to a valid memory
+   location OR the value `0`, which is defined as the NULL sentinel.
+ - `C_FUNCTION_POINTER_OR_ZERO`: This describes an argument which holds a
+   pointer to an executable function, where the function adheres to the
+   standard C ABI for the platform (akin to `extern C` in C++ or `extern "C"
+   fn` in Rust) OR the value `0`, which is defined as the NULL sentinel.
+ - `OPAQUE_GENERIC`: This describes an argument with no restrictions on type.
```

Comment:
I think this would be better as a table (conceptual type, C inspiration, description)

Reply:
Done.